// @nuxthub/core/dist/runtime/ai/server/utils/cloudflare.js
import { createError } from "h3";
import { createCloudflareAPI, handleCloudflareError } from "../../../utils/cloudflare.js";
export function createCloudflareAI(accountId, apiToken) {
  const $api = createCloudflareAPI(accountId, apiToken);
  return {
    aiGatewayLogId: null,
    async run(model, params, _options) {
      const path = `/ai/run/${model}`;
      if (params?.stream) {
        return $api(path, {
          method: "POST",
          body: params,
          // @ts-expect-error - responseType 'stream' is valid but not in ofetch types
          responseType: "stream",
          onResponseError: ({ response }) => handleCloudflareError({ data: response._data, status: response.status }, "Cloudflare AI API error")
        });
      }
      const res = await $api(path, {
        method: "POST",
        body: params,
        onResponseError: ({ response }) => handleCloudflareError({ data: response._data, status: response.status }, "Cloudflare AI API error")
      });
      return res?.result || res;
    },
    async models(params) {
      const res = await $api("/ai/models/search", {
        method: "POST",
        body: params,
        onResponseError: ({ response }) => handleCloudflareError({ data: response._data, status: response.status }, "Cloudflare AI API error")
      });
      return res.result;
    },
    async toMarkdown(_files, _options) {
      throw createError({
        statusCode: 501,
        message: "hubAI().toMarkdown() is not supported when using direct Cloudflare API. Use remote storage or the NuxtHub Admin instead."
      });
    }
  };
}
export function createCloudflareAutoRAG(instance, accountId, apiToken) {
  const $api = createCloudflareAPI(accountId, apiToken, `/autorag/rags/${instance}`);
  return {
    async aiSearch(params) {
      if (params?.stream) {
        return $api("/ai-search", {
          method: "POST",
          body: params,
          // @ts-expect-error - responseType 'stream' is valid but not in ofetch types
          responseType: "stream",
          onResponseError: ({ response }) => handleCloudflareError({ data: response._data, status: response.status }, "Cloudflare AutoRAG API error")
        });
      }
      const res = await $api("/ai-search", {
        method: "POST",
        body: params,
        onResponseError: ({ response }) => handleCloudflareError({ data: response._data, status: response.status }, "Cloudflare AutoRAG API error")
      });
      return res.result;
    },
    async search(params) {
      const res = await $api("/search", {
        method: "POST",
        body: params,
        onResponseError: ({ response }) => handleCloudflareError({ data: response._data, status: response.status }, "Cloudflare AutoRAG API error")
      });
      return res.result;
    }
  };
}
