// @nuxthub/core/dist/runtime/blob/server/utils/cloudflare.js
import { createCloudflareAPI, handleCloudflareError } from "../../../utils/cloudflare.js";
export async function createCloudflareR2Credentials(accountId, apiToken, bucketId, options = {}) {
  const $api = createCloudflareAPI(accountId, apiToken);
  const accessKeyId = await $api(`/tokens/verify`).then((res2) => res2?.result?.id).catch(() => null);
  const res = await $api(`/r2/buckets/${bucketId}/temporary-credentials`, {
    method: "POST",
    body: {
      parentAccessKeyId: accessKeyId,
      ttlSeconds: options.ttl || 900,
      permission: options.permission || "admin-read-write",
      prefixes: options.prefixes,
      objects: options.pathnames
    },
    onResponseError: ({ response }) => handleCloudflareError({ data: response._data, status: response.status }, "Failed to create temporary R2 credentials")
  });
  const result = res?.result || res;
  return {
    accountId,
    bucketName: result.bucketName || bucketId,
    accessKeyId: result.accessKeyId,
    secretAccessKey: result.secretAccessKey,
    sessionToken: result.sessionToken
  };
}
